cranelift/codegen/src/isa/x64/lower.rs

```diff
@@ -2183,41 +2183,93 @@ fn lower_insn_to_regs<C: LowerCtx<I = Inst>>(
-                        RegMem::reg(extend_input_to_reg(ctx, inputs[0], ExtSpec::ZeroExtendTo64));
-                    ctx.emit(Inst::gpr_to_xmm(opcode, src, OperandSize::Size64, dst));
-                }
+                        let src = RegMem::reg(extend_input_to_reg(
+                            ctx,
+                            inputs[0],
+                            ExtSpec::ZeroExtendTo64,
+                        ));
+                        ctx.emit(Inst::gpr_to_xmm(opcode, src, OperandSize::Size64, dst));
+                    }
 
-                types::I64 => {
-                    let src = put_input_in_reg(ctx, inputs[0]);
+                    types::I64 => {
+                        let src = put_input_in_reg(ctx, inputs[0]);
 
-                    let src_copy = ctx.alloc_tmp(RegClass::I64, types::I64);
-                    ctx.emit(Inst::gen_move(src_copy, src, types::I64));
+                        let src_copy = ctx.alloc_tmp(RegClass::I64, types::I64);
+                        ctx.emit(Inst::gen_move(src_copy, src, types::I64));
 
-                    let tmp_gpr1 = ctx.alloc_tmp(RegClass::I64, types::I64);
-                    let tmp_gpr2 = ctx.alloc_tmp(RegClass::I64, types::I64);
-                    ctx.emit(Inst::cvt_u64_to_float_seq(
-                        ty == types::F64,
-                        src_copy,
-                        tmp_gpr1,
-                        tmp_gpr2,
-                        dst,
-                    ));
-                }
+                        let tmp_gpr1 = ctx.alloc_tmp(RegClass::I64, types::I64);
+                        let tmp_gpr2 = ctx.alloc_tmp(RegClass::I64, types::I64);
+                        ctx.emit(Inst::cvt_u64_to_float_seq(
+                            ty == types::F64,
+                            src_copy,
+                            tmp_gpr1,
+                            tmp_gpr2,
+                            dst,
+                        ));
+                    }
+                    _ => panic!("unexpected input type for FcvtFromUint: {:?}", input_ty),
+                };
+            } else {
+                // Converting packed unsigned integers to packed floats requires a few steps.
+                // There is no single instruction lowering for converting unsigned floats but there
+                // is for converted packed signed integers to float (cvtdq2ps). In the steps below
+                // we isolate the upper half (16 bits) and lower half (16 bits) of each lane and
+                // then we convert each half separately using cvtdq2ps meant for signed integers.
+                // In order for this to work for the upper half bits we must shift right by 1
+                // (divide by 2) these bits in order to ensure the most significant bit is 0 not
+                // signed, and then after the conversion we double the value. Finally we add the
+                // converted values where addition will correctly round.
```

Comment:
This is quite clever! A few questions:

- This all assumes 32x4, right? Do we also need to support 64x2 (converting into doubles rather than floats)? If not, can we add an assert (not just a debug assert; we should check in release too)?

- I was quite confused at first as I had thought (from this comment) that this was somehow using a 16-bit conversion rather than a 32-bit conversion with either the upper or lower 16 bits zeroed. I think it would be helpful to add some pseudocode or even ASCII art or something to illustrate what's going on -- for each lane, we basically have:

```
upper := (src & 0xffff0000) >> 1
lower := (src & 0x0000ffff)

dst := float(upper)*2 + float(lower)
```

which makes it a little easier to see the equivalence.

Reply:
> This all assumes 32x4, right? Do we also need to support 64x2 (converting into doubles rather than floats)? If not, can we add an assert (not just a debug assert; we should check in release too)?

Yes, unsigned into float is only 32x4 for WASM. I can add a guard. Also I can add some additional comment/illustration to explain the algorithm better.  👍

Reply:
@cfallin Also, this patch submitted as part of different patch set (branch) without realizing the impact. So patch merged yesterday so I'll address the comments in a separate patch from this one.

Reply:
Ah, OK, no worries and thanks!